File: Python/merge_sort.py
"""
Merge Sort - Online Orders by Delivery Time
Problem: Sort orders by estimated delivery time for priority delivery
Time Complexity: O(n log n)
Space Complexity: O(n)
"""

def merge_sort(arr):
    """Recursive merge sort algorithm"""
    if len(arr) > 1:
        mid = len(arr) // 2
        L = arr[:mid]  # Left half
        R = arr[mid:]  # Right half
        
        # Recursively sort both halves
        merge_sort(L)
        merge_sort(R)
        
        # Merge sorted halves
        i = j = k = 0
        
        # Compare elements and merge
        while i < len(L) and j < len(R):
            if L[i] < R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1
        
        # Copy remaining elements from L
        while i < len(L):
            arr[k] = L[i]
            i += 1
            k += 1
        
        # Copy remaining elements from R
        while j < len(R):
            arr[k] = R[j]
            j += 1
            k += 1

# Driver code
if __name__ == "__main__":
    # Delivery times in minutes
    orders = [45, 10, 30, 60, 25]
    
    print("Unsorted Delivery Times:", orders)
    
    merge_sort(orders)
    
    print("Sorted by Delivery Time (Ascending):", orders)
    print("\nPriority Order:")
    for i, time in enumerate(orders, 1):
        print(f"  {i}. Deliver in {time} minutes")

File: CPP/merge_sort.cpp
/*
 * Merge Sort - Online Orders by Delivery Time
 * Problem: Sort online orders by estimated delivery time (ascending)
 * Time Complexity: O(n log n)
 * Space Complexity: O(n)
 */

#include <iostream>
using namespace std;

void merge(int a[], int l, int m, int r) {
    int n1 = m - l + 1;
    int n2 = r - m;
    
    // Create temporary arrays
    int L[n1], R[n2];
    
    // Copy data to temp arrays
    for(int i = 0; i < n1; i++)
        L[i] = a[l + i];
    for(int j = 0; j < n2; j++)
        R[j] = a[m + 1 + j];
    
    // Merge the temp arrays back
    int i = 0, j = 0, k = l;
    while(i < n1 && j < n2) {
        if(L[i] <= R[j]) {
            a[k] = L[i];
            i++;
        } else {
            a[k] = R[j];
            j++;
        }
        k++;
    }
    
    // Copy remaining elements
    while(i < n1) {
        a[k] = L[i];
        i++;
        k++;
    }
    while(j < n2) {
        a[k] = R[j];
        j++;
        k++;
    }
}

void mergeSort(int a[], int l, int r) {
    if(l < r) {
        int m = (l + r) / 2;
        
        // Sort first and second halves
        mergeSort(a, l, m);
        mergeSort(a, m + 1, r);
        
        // Merge the sorted halves
        merge(a, l, m, r);
    }
}

int main() {
    int time[] = {45, 10, 30, 60, 25};
    int n = 5;
    
    cout << "Delivery times (unsorted): ";
    for(int i = 0; i < n; i++)
        cout << time[i] << " ";
    
    mergeSort(time, 0, n - 1);
    
    cout << "\nSorted by delivery time (ascending): ";
    for(int i = 0; i < n; i++)
        cout << time[i] << " ";
    cout << endl;
    
    return 0;
}
